### ISSUE TITLE
Logarithm of zero: why does `math.log(0)` not follow IEEE 754 recommendations?

### ISSUE THREAD

**Issue by randolf-scholz** [python/cpython#94621]
[IEEE Standard for Binary Floating-Point Arithmetic](https://ieeexplore.ieee.org/document/30711) recommends that `logb(0)` should return negative infinity. (cf. Annex A Recommended Functions and Predicates). In Python however, we currently (3.10.5) have:

- `math.log(0)` raises `ValueError`

On the other hand, `math.log(math.inf)` returns `math.inf`. It seems arbitrary to support this limit case but not the other. ~~The [documentation](https://docs.python.org/3/library/math.html) is silent on this issue as well.~~ It is mentioned at the very end but not within the scope of the log-function itself.

Relevant lines: can we simply change the `<=` to `<` and give a RunTimeWanring like `numpy` in the case of `0`?

https://github.com/python/cpython/blob/23c9febdc6620c2ec5b6119d9016a8c92c25f350/Modules/mathmodule.c#L2336-L2341

**Comment by stevendaprano** [comment]
Unfortunately you have linked to a non-free standard, which means its effectively unreadable.

What would be the purpose of the RuntimeWarning if we returned -inf?

**Comment by randolf-scholz** [comment]
@stevendaprano [IEEE-754](https://en.wikipedia.org/wiki/IEEE_754), is well-known, there are already some 200 issues in this repository referencing it. Whether to yield a warning when `log(0)`  is computed is debatable. Numpy does it by default, but I am not I'm not convicted in this matter.

**Comment by mdickinson** [comment]
Thanks for the report. The behaviour is deliberate, and is indeed based on IEEE 754 (or more accurately, C99 Annex F, but it comes to the same thing).

CPython's philosophy on this is set out at the top of the mathmodule source: https://github.com/python/cpython/blob/bde06e1b8381f140b296a397ddd1deb1c784ff8e/Modules/mathmodule.c#L9-L53.
tl;dr: Python shouldn't produce infinities from finite inputs, and shouldn't produce nans from non-NaN inputs.

The behaviours were originally based on C99 Annex F, since IEEE 754 didn't cover transcendental functions until the 2008 revision. The two standards are almost completely compatible in their coverage of special cases, though.
In brief: where IEEE 754 / C99 signals divide-by-zero or invalid, Python aims to raise `ValueError`; where IEEE 754 signals overflow, Python aims to raise `OverflowError`. No attempt is made to do anything with inexact or underflow signals.

Note that the asymmetry you're observing exists in IEEE 754 too: `log(0)` signals divide-by-zero, `log(infty)` returns infinity and does not signal.

**Comment by randolf-scholz** [comment]
Ok that makes sense, I guess the question then becomes whether, following section 8 "Alternate exception handling attributes" of the 2019 revision of IEEE 754, to include a switch:

> Language standards should define, and require implementations to provide, means for the user to associate
alternate exception handling attributes with blocks (see 4.1)

For example, `numpy` has [`numpy.seterr`](https://numpy.org/doc/stable/reference/generated/numpy.seterr.html), which allows the user to decide whether to raise, warn, or quiet the exceptions specified in IEEE 754.

**Comment by mdickinson** [comment]
> Would it make sense to open a feature request for a similar functionality in pure python?

It's certainly an option. It would be a fairly significant change, with quite a lot of moving parts to consider - you'd be introducing global (or at least thread-local) state, so there are questions about how and where that state is stored, how it interacts with things like subinterpreters and PEP 567-style contexts, and what mechanisms should be provided to the user for state modifications (e.g., context managers?). It would likely need a core developer champion to help push it through. But we at least have NumPy to draw on as prior art. You might want to start by posting on the python-ideas mailing list or the Ideas channel at discuss.python.org to gauge whether there's community interest in the idea.
